File: platform/rt-sandbox-ui/src/runtime/livePreflight.ts
/** Live Gazebo host preflight mirror (rt_live_runtime_preflight_v1). */

export type LivePreflightChecks = {
  ros2_available: boolean;
  gz_available: boolean;
  rt_sandbox_gz_available: boolean;
};

export type LivePreflightResult = {
  schema?: string;
  ok: boolean;
  checks: LivePreflightChecks;
  blockers: string[];
  message?: string;
};

export type LivePreflightRow = {
  id: keyof LivePreflightChecks;
  label: string;
  ok: boolean;
};

const CHECK_LABELS: Record<keyof LivePreflightChecks, string> = {
  ros2_available: "ros2 on PATH",
  gz_available: "gz (Gazebo Sim) on PATH",
  rt_sandbox_gz_available: "rt_sandbox_gz ROS package",
};

export function parseLivePreflight(value: unknown): LivePreflightResult | null {
  if (!value || typeof value !== "object") return null;
  const raw = value as Record<string, unknown>;
  const checksRaw = raw.checks;
  if (!checksRaw || typeof checksRaw !== "object") return null;
  const checksIn = checksRaw as Record<string, unknown>;
  const checks: LivePreflightChecks = {
    ros2_available: checksIn.ros2_available === true,
    gz_available: checksIn.gz_available === true,
    rt_sandbox_gz_available: checksIn.rt_sandbox_gz_available === true,
  };
  const blockers = Array.isArray(raw.blockers)
    ? raw.blockers.filter((item): item is string => typeof item === "string")
    : [];
  return {
    schema: typeof raw.schema === "string" ? raw.schema : undefined,
    ok: raw.ok === true,
    checks,
    blockers,
    message: typeof raw.message === "string" ? raw.message : undefined,
  };
}

export function livePreflightRows(result: LivePreflightResult | null): LivePreflightRow[] {
  if (!result) {
    return (Object.keys(CHECK_LABELS) as Array<keyof LivePreflightChecks>).map((id) => ({
      id,
      label: CHECK_LABELS[id],
      ok: false,
    }));
  }
  return (Object.keys(CHECK_LABELS) as Array<keyof LivePreflightChecks>).map((id) => ({
    id,
    label: CHECK_LABELS[id],
    ok: result.checks[id],
  }));
}

export function livePreflightSummary(result: LivePreflightResult | null): string {
  if (!result) return "Preflight not checked yet.";
  if (result.ok) return "Live runtime preflight passed.";
  if (result.blockers.length > 0) return result.blockers.join("; ");
  return result.message ?? "Live runtime unavailable.";
}

File: platform/rt-sandbox-bridge/rt_sandbox/live_preflight.py
"""Live Gazebo runtime preflight checks (Web ↔ Gazebo Step 2).

Read-only host probes — bridge only; no auto-install.
"""

from __future__ import annotations

import shutil
import subprocess
from typing import Any


def _ros_pkg_available(package: str) -> bool:
    if shutil.which("ros2") is None:
        return False
    try:
        proc = subprocess.run(
            ["ros2", "pkg", "prefix", package],
            capture_output=True,
            text=True,
            timeout=10,
        )
    except (OSError, subprocess.TimeoutExpired):
        return False
    return proc.returncode == 0


def check_live_runtime_preflight() -> dict[str, Any]:
    """Return schema rt_live_runtime_preflight_v1 — fail-closed when any check fails."""
    ros2_available = shutil.which("ros2") is not None
    gz_available = shutil.which("gz") is not None
    rt_sandbox_gz_available = _ros_pkg_available("rt_sandbox_gz")

    checks = {
        "ros2_available": ros2_available,
        "gz_available": gz_available,
        "rt_sandbox_gz_available": rt_sandbox_gz_available,
    }
    blockers: list[str] = []
    if not ros2_available:
        blockers.append("ros2 not found on PATH")
    if not gz_available:
        blockers.append("gz (Gazebo Sim) not found on PATH")
    if not rt_sandbox_gz_available:
        blockers.append("rt_sandbox_gz ROS package unavailable (source install/setup.bash)")

    ok = not blockers
    message = "live runtime ready" if ok else "; ".join(blockers)
    return {
        "schema": "rt_live_runtime_preflight_v1",
        "ok": ok,
        "checks": checks,
        "blockers": blockers,
        "message": message,
    }
